src/coreclr/jit/lower.cpp

```diff
@@ -5447,6 +5452,62 @@ GenTree* Lowering::LowerAdd(GenTreeOp* node)
             JITDUMP("Remove [%06u], [%06u]\n", op2->gtTreeID, node->gtTreeID);
             return next;
         }
+#ifdef TARGET_ARM64
+        if (BlockRange().TryGetUse(node, &use))
+        {
+            if (comp->opts.OptimizationEnabled() && varTypeIsIntegral(node) &&
+                !(node->gtFlags & GTF_SET_FLAGS) && !node->gtOverflow() &&
+                !node->isContained())
+            {
+                GenTree* mul       = nullptr;
+                GenTree* c         = nullptr;
+                if (op1->OperIs(GT_MUL))
+                {
+                    mul = op1;
+                    c   = op2;
+                }
+                else
+                {
+                    mul = op2;
+                    c   = op1;
+                }
+
+                if (mul->OperIs(GT_MUL) && !(mul->gtFlags & GTF_SET_FLAGS) && varTypeIsIntegral(mul) &&
+                    !mul->gtOverflow() && !mul->isContained() && !c->isContained())
+                {
+                    GenTree* a = mul->gtGetOp1();
+                    GenTree* b = mul->gtGetOp2();
+
+                    // Transform "-a * b + c" to "c - a * b"
+                    if (a->OperIs(GT_NEG) && !(a->gtFlags & GTF_SET_FLAGS) && !b->OperIs(GT_NEG) && !a->isContained() &&
+                        !a->gtGetOp1()->isContained())
+                    {
+                        mul->AsOp()->gtOp1 = a->gtGetOp1();
+                        BlockRange().Remove(a);
+                        node->gtOp1 = c;
+                        node->gtOp2 = mul;
+                        node->ChangeOper(GT_SUB);
+                    }
+                    // Transform "a * -b + c" to "c - a * b"
+                    else if (b->OperIs(GT_NEG) && !(b->gtFlags & GTF_SET_FLAGS) && !a->OperIs(GT_NEG) &&
+                             !b->isContained() && !b->gtGetOp1()->isContained())
+                    {
+                        mul->AsOp()->gtOp2 = b->gtGetOp1();
+                        BlockRange().Remove(b);
+                        node->gtOp1 = c;
+                        node->gtOp2 = mul;
+                        node->ChangeOper(GT_SUB);
+                    }
+                    // Transform "a * b + c" to "c + a * b"
+                    else if (op1->OperIs(GT_MUL))
+                    {
+                        node->gtOp1 = c;
+                        node->gtOp2 = mul;
+                    }
```

Comment:
The reordering that is happening here makes me a little uneasy. Since locals are used at their user, it is now technically possible that this reorders the read of a local with a write of the same local.

Reply:
If anything the problem will happen during containment, so this code at least should be ok. And the containment checks are preexisting, so I suppose it is fine in practice.

Reply:
I know what you mean, I felt the same way. I spoke to @tannergooding about it.

As far as I know, the re-ordering of the tree itself doesn't change the linear execution unless containment happens - correct me if I'm wrong here... So I think it should be ok.